spec/index.bs

```diff
@@ -1403,6 +1427,97 @@ but which the user agent has not been updated to recognize yet. If the [=set ent
 </div>
 </div>
 
+#### <dfn interface>GPUAdapterInfo</dfn> #### {#gpu-adapterinfo}
+
+{{GPUAdapterInfo}} exposes various identifying information about an adapter.
+
+Note: None of the values in {{GPUAdapterInfo}} are guaranteed to be populated. It is at the user
+agent's discretion which values to reveal, and it is likely that on some devices none of the values
+will be populated. As such applications should not be dependent on any values from {{GPUAdapterInfo}}
+to function.
+
+<script type=idl>
+[Exposed=(Window, DedicatedWorker), SecureContext]
+interface GPUAdapterInfo {
+  readonly attribute DOMString vendor;
+  readonly attribute DOMString architecture;
+  readonly attribute long deviceId;
+  readonly attribute DOMString description;
+};
+</script>
+
+{{GPUAdapterInfo}} has the following attributes:
+
+<dl dfn-type=attribute dfn-for=GPUAdapterInfo>
+    : <dfn>vendor</dfn>
+    ::
+        The name of the vendor of the [=adapter=], if avaliable. Empty string otherwise.
+
+    : <dfn>architecture</dfn>
+    ::
+        The name of the family or class of GPUs the [=adapter=] belongs to, if avaliable. Empty
+        string otherwise.
+
+    : <dfn>deviceId</dfn>
```

Comment:
Exposing PCI device IDs to the web sounds like a terrible idea

Reply:
I understand that sentiment, but I'm not sure what a useful alternative would be.

This value is exposed directly by Vulkan and D3D12 and named "deviceId" by both APIs. As far as I can tell Metal has no direct equivalent, but it should be able to populate the `architecture` attribute more accurately than other systems, assuming that the GPU family is a reasonable value to surface there. (Please let me know if you feel it's not!)

We could surface it as a string instead, but the implication there is that we would need to maintain a lookup table of every known vendor/device ID pairing to the product name, which is a large ongoing engineering time commitment that I don't want to require if I can help it. (I'm already uncomfortable with the mapping that `architecture` implies on non-Metal systems.)

On the other hand, if we don't expose anything more detailed than `architecture` on non-Metal systems there will be a strong motivation for developers to attempt to parse the `description` field to identify the specific device, which I definitely don't want to encourage.

As such, while I'm not a big fan of throwing an opaque number at the developer considering how we intend for them to use it (identifying buggy devices) it fulfills the need. Worth noting that I think Chrome would always require some form of user consent before unmasking this value.

Reply:
Had a conversation with @kenrussell today in which he pointed out that the ANGLE team has had previous experience trying to expose a value like the `deviceId` proposed here, and suggested that we use a string for that purpose instead. Specifically, he pointed out that one some devices the ANGLE team needed to report a 64 bit number. If we use a string here then there's a lot more flexibility in what can be reported.

I would still expect some platforms/UAs to report a PCI device ID here, probably just formatted as a hex string.

I also took the liberty of renaming the value to simply "device" since it's not going to be strictly numeric. Still have some mild concerns about the terminology conflict with `GPUDevice` and am open to alternatives, but I think the context will probably clear up any ambiguity about it's use.